service-coordtransform/src/main/java/fi/nls/paikkatietoikkuna/coordtransform/CoordinateTransformationActionHandler.java

```diff
@@ -83,20 +197,181 @@ public void handleAction(ActionParameters params) throws ActionException {
+        String coordSeparator = coordinateSeparators.get(sourceOptions.getCoordinateSeparator());
+        if (sourceOptions.isAxisFlip()){
+            xIndex = 1;
+            yIndex = 0;
+        }
+        if (sourceOptions.isPrefixId()){
+            xIndex++;
+            yIndex++;
+            zIndex++;
+            coordDimension++;
+        }
+        boolean replaceCommas = false;
+        if (sourceOptions.getDecimalSeparator()==','){
+            replaceCommas = true;
+        }
+        String unit = sourceOptions.getUnit();
+        boolean transformUnit = false;
+        if (unit != null && !unit.equals(DEGREE)){
+            transformUnit = true;
+        }
+        double x,y,z;
+        try{
+            BufferedReader br = new BufferedReader(
+                    new InputStreamReader(file.getInputStream()));
+            //skip row and store row as header row
+            for (int i = 0 ; i < headerLineCount && (line=br.readLine())!=null ;i++) {
+                sourceOptions.addHeaderRow(line);
+            }
+            while ((line=br.readLine())!=null) {
+                /* Now coordinate separator comes from frontend
+                //try to get coordinate separator from first coordinate line
+                if(coordSeparator==null){
+                    coordSeparator = CoordTransService.getCoordSeparator(line, dimension, sourceOptions.isPrefixId());
+                    sourceOptions.setCoordinateSeparator(coordSeparator);
+                }*/
+                //skip empty lines
+                if (line.trim().isEmpty()){
+                    continue;
+                }
+                //replace commas
+                if (replaceCommas){
+                    line = line.replace(',','.');
+                }
+                coords = line.split(coordSeparator);
+                if (transformUnit){
+                    x = CoordTransService.transformUnitToDegree (coords[xIndex], unit);
+                    y = CoordTransService.transformUnitToDegree (coords[yIndex], unit);
+                } else {
+                    x = Double.valueOf(coords[xIndex]);
+                    y = Double.valueOf(coords[yIndex]);
+                }
+                if (dimension == 3){
+                    z = Double.valueOf(coords[zIndex]);
+                    coordinates.add(new Coordinate(x, y, z));
+                }else if (addZeroes == true){
+                    coordinates.add(new Coordinate(x, y, 0));
+                } else {
+                    coordinates.add(new Coordinate(x, y));
+                }
+                if (sourceOptions.isPrefixId()){
+                    sourceOptions.addId(coords[0]);
+                }
+                if (storeLineEnds == true){
+                    String lineEnd = "";
+                    //add coordSeparator back if lineEnding string is slitted (e.g. coordSeparator is " ")
+                    for (int i = coordDimension; i < coords.length; i++){
```

Comment:
Coordinate line can be id,x,y,z,lineEnding where lineEnding is all data after coordinates. If lineEnding contains coordSeparator chars then it is splitted also. Might be better to use limit with split?

Reply:
This would be easy to solve if we replaced the whole split() based logic by a bunch of indexOf() and substring() calls. That's overall less pretty though.